platform/pwl.c

```diff
@@ -39,3 +61,775 @@ pwl_display_destroy (PwlDisplay *self)
+            wl_data.metrics[i].name = 0;
+            g_debug ("Removed output %i\n", name);
+            break;
+        }
+    }
+}
+#endif /* HAVE_DEVICE_SCALING */
+
+static const struct wl_registry_listener registry_listener = {
+    .global = registry_global,
+#if HAVE_DEVICE_SCALING
+    .global_remove = registry_global_remove
+#endif /* HAVE_DEVICE_SCALING */
+};
+
+gboolean init_egl (GError **error)
+{
+    g_debug ("Initializing EGL...");
+
+    egl_data.display = eglGetDisplay ((EGLNativeDisplayType) wl_data.display);
+    if (egl_data.display == EGL_NO_DISPLAY) {
+        // TODO: ERR_EGL (error, "Could not open EGL display");
+        return FALSE;
+    }
+
+    EGLint major, minor;
+    if (!eglInitialize (egl_data.display, &major, &minor)) {
+        // TODO: ERR_EGL (error, "Could not initialize  EGL");
+        clear_egl ();
+        return FALSE;
+    }
+    g_info ("EGL version %d.%d initialized.", major, minor);
+
+    if (!eglBindAPI (EGL_OPENGL_ES_API)) {
+        // TODO: ERR_EGL (error, "Could not bind OpenGL ES API to EGL");
+        clear_egl ();
+        return FALSE;
+    }
+
+    static const EGLint context_attribs[] = {
+        EGL_CONTEXT_CLIENT_VERSION, 2,
+        EGL_NONE
+    };
+
+    static const EGLint config_attribs[] = {
+        EGL_SURFACE_TYPE, EGL_WINDOW_BIT,
+        EGL_RED_SIZE,     8,
+        EGL_GREEN_SIZE,   8,
+        EGL_BLUE_SIZE,    8,
+        EGL_ALPHA_SIZE,   0,
+        EGL_DEPTH_SIZE,   0,
+        EGL_STENCIL_SIZE, 0,
+        EGL_RENDERABLE_TYPE, EGL_OPENGL_ES2_BIT,
+        EGL_SAMPLES, 0,
+        EGL_NONE
+    };
+
+    EGLint num_configs;
+    if (!eglChooseConfig (egl_data.display,
+                          config_attribs,
+                          &egl_data.egl_config,
+                          1,
+                          &num_configs)) {
+        // TODO: ERR_EGL (error, "Could not find a suitable EGL configuration");
+        clear_egl ();
+        return FALSE;
+    }
+    g_assert (num_configs > 0);
+
+    egl_data.context = eglCreateContext (egl_data.display,
+                                         egl_data.egl_config,
+                                         EGL_NO_CONTEXT,
+                                         context_attribs);
+    if (egl_data.context == EGL_NO_CONTEXT) {
+        // TODO: ERR_EGL (error, "Could not create EGL context");
+        clear_egl ();
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+void clear_egl (void)
+{
+    if (egl_data.display != EGL_NO_DISPLAY) {
+        if (egl_data.context != EGL_NO_CONTEXT) {
+            eglDestroyContext (egl_data.display, egl_data.context);
+        }
+        eglTerminate (egl_data.display);
+        egl_data.context = EGL_NO_CONTEXT;
+        egl_data.display = EGL_NO_DISPLAY;
+    }
+    eglReleaseThread ();
+}
+
+
+gboolean init_wayland (GError **error)
+{
+    g_debug ("Initializing Wayland...");
+
+    if (!(wl_data.display = wl_display_connect (NULL))) {
```

Comment:
This is creating an additional Wayland connection, with the other being created by `pwl_display_connect()`. We definitely not want to have more than one connection.

Note that `cog_fdo_module_is_supported()` creates a `PwlDisplay` when the GIO module is loaded and the function called to check whether the module is usable. We want to reuse tha `wl_display` contained in `PwlDisplay` for all the Wayland API uses that need a display.

If you need the `wl_display*` from outside `pwl.c`, we could have an accessor like this:

```c
struct wl_display* pwl_display_get_wl_display(const PwlDisplay *self)
{
    g_return_val_if_fail (self, NULL);
   return self->display;
}
```